csvqb/csvqb/models/cube/csvqb/components/dimension.py

```diff
@@ -2,96 +2,77 @@
 Dimensions
 ----------
 """
+from dataclasses import dataclass, field
 from typing import Optional, List
-from abc import ABC
+from abc import ABC, abstractmethod
 
-import pandas as pd
-
-from csvqb.utils.uri import uri_safe
+from csvqb.models.uriidentifiable import UriIdentifiable
 from .datastructuredefinition import ColumnarQbDataStructureDefinition
 from .codelist import QbCodeList, NewQbCodeList
 from csvqb.models.validationerror import ValidationError
 from csvqb.inputs import PandasDataTypes
 from ..catalog import CatalogMetadata
 
 
+@dataclass
 class QbDimension(ColumnarQbDataStructureDefinition, ABC):
-    def __init__(self, range_uri: Optional[str]):
-        self.range_uri: Optional[str] = range_uri
-
+    @property
+    @abstractmethod
+    def range_uri(self) -> Optional[str]:
```

Comment:
Thanks for your explaination about abstract properties, and how you're using it to get around required properties declared after a child class' optional properties.